scripts/west_commands/ncs_commands.py

```diff
@@ -188,14 +189,22 @@ def print_loot(self, name, project, z_project, args):
 
         log.banner(msg)
 
+        nrev = project.revision
+        if re.fullmatch(r"pull/\d+/head", nrev):
+            # As a special case, match GitHub's magic pull request
+            # ref space.
+            log.small_banner('fetching latest version of ' + nrev)
+            project.git('fetch {url} ' + nrev)
```

Comment:
Some thoughts crossing my mind when looking at this.
This is placed here, because a fetch is done at this location, but it covers a more fundamental question about `ncs-loot`.

`git fetch` is not done anywhere in the `ncs-loot` except here, which mean that a pre-requisite for running `west ncs-loot` must be that `west update` has already been executed.
Otherwise existing code wouldn't work, cause the SHA referenced in the `west.yml` might not be in local repo.

So current code must have the following assumption:
1) User has already executed `west update`

Now, when user did `west update`, then the `pull/<no>/head` is already available as `FETCH_HEAD`, so no need to do a second `git fetch`, unless there is a risk user did a manual `git fetch` after `west update`, which leads to second assumption.
That is, if we rely on current `FETCH_HEAD == FETCH_HEAD from west update`

assumption 2) User does not do a manual `git fetch` after west update.

Now, assumption 1) and 2) may be equally correct or equally wrong, but in either case, I would say that for `west ncs-loot` then:
a) `ncs-loot` doesn't need to do a `git fetch` (risky imho)
b) always do a `git fetch` to ensure the SHAs are always present regardless of whether a SHA or `pull/<no>/head` revision is given in manifest file.

so I will actually suggest `west ncs-loot` always does a `git fetch`.

`git fetch` is already available as (private) method `_fetch` in `project.py`, and because of that I would suggest to expose this as a `class Fetch` directly in `projects.py`, similar to the `class Update` being available.

And then the test for rev being in the form of `pull/<no>/head` will then decide if `FETCH_HEAD` is used for the `cat-file` command.

Reply:
Thanks for the thoughtful review.

In general, I would like these commands to work without a network connection whenever possible. That is why `west ncs-compare` prompts the user to fetch if a ref is not available.

I wrote this patch as a special case because it is a bit special -- normally, zephyr/west.yml and ncs/west.yml only contain SHAs as revisions, so once you've fetched the upstream and downstream master branches, you've got a copy of those SHAs 'forever', without needing to make any assumptions about what FETCH_HEAD points to.

However, when we are inspecting pull requests with `pull/*/head` as revisions, that's not the case. 

One way to keep your local refs in sync with the `pull/*/head` refs on the remote is to have something like this in your .git/config:

```
[remote "origin"]
	url = https://github.com/NordicPlayground/fw-nrfconnect-zephyr
        [... other fetch refspecs go here...]
	fetch = +refs/pull/*/head:refs/pull/*/head
```

Then a `git fetch origin` will sync up the remote PRs, so `west ncs-loot zephyr` will do the right thing offline without having to fetch anything, since a local `pull/N/head` will be available.

But that's a bit messy, and anyway, dealing with refspecs is getting into fairly advanced Git.

My goal here is to make it easier on other NCS maintainers who are reviewing pull requests to just get the latest info when they run `west ncs-loot`. (Remember that this command is for internal use by maintainers.)

I guess one potential compromise would be for me to print a line like this when we can't resolve the ref:

```
can't get loot; please run:
    git fetch {url} +refs/pull/N/head:refs/pull/N/head
```

Where N is a concrete number.

Thoughts?

Reply:
> `git fetch` is already available as (private) method `_fetch` in `project.py`, and because of that I would suggest to expose this as a `class Fetch` directly in `projects.py`, similar to the `class Update` being available.

Neither of these is "available"; that class is not API.

The API is here: https://docs.zephyrproject.org/latest/guides/west/west-apis.html#west-apis

Anything not explicitly listed there is to be considered an implementation detail. In particular, none of the built-in commands are API.

Reply:
> > `git fetch` is already available as (private) method `_fetch` in `project.py`, and because of that I would suggest to expose this as a `class Fetch` directly in `projects.py`, similar to the `class Update` being available.
> 
> Neither of these is "available"; that class is not API.
> 
> The API is here: https://docs.zephyrproject.org/latest/guides/west/west-apis.html#west-apis
> 
> Anything not explicitly listed there is to be considered an implementation detail. In particular, none of the built-in commands are API.

Ahh, you're right.
I was mixing up the API available to extension commands with the `west` built-in commands available to users.

Follow up question, should we consider to make the built-in commands available also as API ?
Like, could it make sense for an extension command to have clean access to `list` and `diff` as example.

I know this is sidetracking this PR a little, but worth a thought.
Although, there are also calls that should not be done by an extension command, like `ForAll`, so i'm also a little reluctant myself to open `project.py` to extension commands.

I have a small idea I want to try, and then post an update later.

Reply:
Made: https://github.com/zephyrproject-rtos/west/pull/331

This will keep revisions locally under `refs/west/` refs space.
So if one uses `pull/<no>/head` or `refs/heads/<branch>` instead of a SHA, then such a revision will be locally available as: `refs/west/pull/<no>/head` and `refs/west/<branch>`.

Thus your code can do:
`cat-file -e refs/west/<revision>` when using non SHA revisions.

And if the revision is not locally avaiable, then ask user to do a `git fetch` as you already do.

Note, this will now work for any non SHA revision.
Where the current suggestion is only working for PRs, and not if user specifies: `refs/heads/<branch>` as revision.

Reply:
and of course, he can also use revision as: `<branch>` and that will go to `refs/west/<branch>`, which is identical to current `git fetch` behavior.

So we make no assumption on the pattern he uses for non SHA revisions, we handle them all ;)

Reply:
> I guess one potential compromise would be for me to print a line like this when we can't resolve the ref:
>
> `  can't get loot; please run:`
> `     git fetch {url} +refs/pull/N/head:refs/pull/N/head`

Should we consider a `west fetch` command ?

Reply:
> Follow up question, should we consider to make the built-in commands available also as API ?
> Like, could it make sense for an extension command to have clean access to `list` and `diff` as example.

I don't think so. We can certainly move useful code in the built-ins to helper methods in the WestCommand API, but the built-in command implementations should remain outside of the API.

> Made: [zephyrproject-rtos/west#331](https://github.com/zephyrproject-rtos/west/pull/331)

Reviewed over there, thanks :)

> Should we consider a `west fetch` command ?

Perhaps depending on the outcome of #331. As discussed there, it's technically breaking even more rules :)

Reply:
> > Should we consider a `west fetch` command ?
> 
> Perhaps depending on the outcome of #331. As discussed there, it's technically breaking even more rules :)

@tejlmand: I now agree with you about this.

I think the idea of fetching `{revision}` directly into manifest-rev as I proposed in zephyrproject-rtos/west#331 today, plus `west fetch`, would make this pull request unnecessary, since `west fetch` with a `pull/xxx/head` would update `manifest-rev` to point directly at that pull request. Then `ncs-loot` could compare upstream zephyr/west.yml's revision against the NCS `manifest-rev` pointer after the user runs `west fetch`. Done and done :).

I'll close this now. Thanks for your patience and great work in west 331.